internal/controllers/objectsetphases/phase_reconciler.go

```diff
@@ -0,0 +1,482 @@
+	if len(desiredObj.GetNamespace()) == 0 {
+		desiredObj.SetNamespace(
+			owner.GetNamespace())
+	}
+
+	// Set cache label
+	labels := desiredObj.GetLabels()
+	if labels == nil {
+		labels = map[string]string{}
+	}
+	labels[controllers.DynamicCacheLabel] = "True"
+	desiredObj.SetLabels(labels)
+
+	// Set owner reference
+	if err := r.ownerStrategy.SetControllerReference(owner, desiredObj); err != nil {
+		return nil, err
+	}
+	return desiredObj, nil
+}
+
+type CommonObjectPhaseError struct {
+	OwnerKey, ObjectKey client.ObjectKey
+	OwnerGVK, ObjectGVK schema.GroupVersionKind
+}
+
+// This error is returned when a Phase contains objects
+// that are not owned by a previous revision.
+// Previous revisions of an Phase have to be declared in .spec.previousRevisions.
+type ObjectNotOwnedByPreviousRevisionError struct {
+	CommonObjectPhaseError
+}
+
+func (e ObjectNotOwnedByPreviousRevisionError) Error() string {
+	return fmt.Sprintf("refusing adoption, object %s %s not owned by previous revision", e.ObjectGVK, e.ObjectKey)
+}
+
+// This error is returned when a Phase tries to adopt an object
+// where the revision number is not increasing.
+type RevisionCollisionError struct {
+	CommonObjectPhaseError
+}
+
+func (e RevisionCollisionError) Error() string {
+	return fmt.Sprintf("refusing adoption, revision collision on %s %s", e.ObjectGVK, e.ObjectKey)
+}
+
+func (r *PhaseReconciler) reconcileObject(
+	ctx context.Context, owner PhaseObjectOwner,
+	desiredObj *unstructured.Unstructured,
+) (actualObj *unstructured.Unstructured, err error) {
+	objKey := client.ObjectKeyFromObject(desiredObj)
+	currentObj := desiredObj.DeepCopy()
+	err = r.dynamicCache.Get(ctx, objKey, currentObj)
+	if err != nil && !errors.IsNotFound(err) {
+		return nil, fmt.Errorf("getting %s: %w", desiredObj.GroupVersionKind(), err)
+	}
+	if errors.IsNotFound(err) {
+		// The object is not yet present on the cluster,
+		// just create it using desired state!
+		err := r.writer.Create(ctx, desiredObj)
+		if err != nil {
+			return nil, fmt.Errorf("creating: %w", err)
+		}
+		return desiredObj, nil
+	}
+
+	// An object already exists - this is the complicated part.
+
+	// Keep a copy of the object on the cluster for comparison.
+	// UpdatedObj will be changed according to desiredObj.
+	updatedObj := currentObj.DeepCopy()
+
+	// Check if we can even work on this object or need to adopt it.
+	needsAdoption, err := r.adoptionChecker.Check(ctx, owner, currentObj)
+	if err != nil {
+		return nil, err
+	}
+
+	// Take over object ownership by patching metadata.
+	if needsAdoption {
+		log := logr.FromContextOrDiscard(ctx)
+		log.Info("adopting object",
+			"OwnerKey", client.ObjectKeyFromObject(owner.ClientObject()),
+			"OwnerGVK", owner.ClientObject().GetObjectKind().GroupVersionKind(),
+			"ObjectKey", client.ObjectKeyFromObject(desiredObj),
+			"ObjectGVK", desiredObj.GetObjectKind().GroupVersionKind())
+		controllers.SetObjectRevision(updatedObj, owner.GetStatusRevision())
```

Comment:
Is this correct? If the revision number of the owner is less than the revision number of the object, isn't it possible to break the monotonically increase nature of revisions?

Reply:
It is correct - as far as I am able to tell, all the checks are being done in the `defaultAdoptionChecker`.
Only when `owner.GetStatusRevision()` is greater than current Revision is needAdoption set to true.